enterprise/internal/database/perms_store.go

```diff
@@ -696,42 +696,56 @@ func (s *PermsStore) loadExistingUserPendingPermissionsBatch(ctx context.Context
 	return bindIDsToIDs, nil
 }
 
+// upsertUserPendingPermissionsBatchQuery generates a query for upserting the provided
+// external service accounts into `user_pending_permissions`.
 func upsertUserPendingPermissionsBatchQuery(
 	accounts *extsvc.Accounts,
 	p *authz.RepoPermissions,
 ) (*sqlf.Query, error) {
+	// Above ~10,000 accounts (10,000 * 6 fields each = 60,000 parameters), we can run
+	// into the Postgres parameter limit inserting with VALUES. Instead, we pass in fields
+	// as arrays, where each array only counts for a single parameter.
 	const format = `
 -- source: enterprise/internal/database/perms_store.go:upsertUserPendingPermissionsBatchQuery
 INSERT INTO user_pending_permissions
   (service_type, service_id, bind_id, permission, object_type, updated_at)
-VALUES
-  %s
+  (
+	SELECT *
+	FROM UNNEST(%s::TEXT[], %s::TEXT[], %s::TEXT[], %s::TEXT[], %s::TEXT[], %s::TIMESTAMPTZ[])
```

Comment:
We can actually optimize this because we have most of them as "static" values:

```suggestion
SELECT %s::TEXT, %s::TEXT, UNNEST(%s::TEXT[]), %s::TEXT, %s::TEXT, UNNEST(%s::TEXT[])
```

<img width="664" alt="CleanShot 2021-09-15 at 21 29 04@2x" src="https://user-images.githubusercontent.com/2946214/133442214-c5da33f1-1fe8-4a95-89ce-531cdb11bf12.png">

Reply:
Right, nice! 6fcfbdbffda4242810c0ac538bc17c42fe1f8604